src/controllers/queryRunner.ts

```diff
@@ -355,63 +356,83 @@ export default class QueryRunner {
             return async () => {
                 const result = await this.getRows(range.fromRow, range.toRow - range.fromRow + 1, batchId, resultId);
-                if (this.shouldIncludeHeaders(includeHeaders)) {
-                    let columnHeaders = this.getColumnHeaders(batchId, resultId, range);
-                    if (columnHeaders !== undefined) {
-                        copyString += columnHeaders.join('\t') + os.EOL;
-                    }
-                }
-
-                // Iterate over the rows to paste into the copy string
-                for (let rowIndex: number = 0; rowIndex < result.resultSubset.rows.length; rowIndex++) {
-                    let row = result.resultSubset.rows[rowIndex];
-                    let cellObjects = row.slice(range.fromCell, (range.toCell + 1));
-                    // Remove newlines if requested
-                    let cells = this.shouldRemoveNewLines()
-                        ? cellObjects.map(x => this.removeNewLines(x.displayValue))
-                        : cellObjects.map(x => x.displayValue);
-                    copyString += cells.join('\t');
-                    if (rowIndex < result.resultSubset.rows.length - 1) {
-                        copyString += os.EOL;
+                for (let row of result.resultSubset.rows) {
+                    let rowNumber = row[0].rowId + range.fromRow;
+                    if (rowIdToSelectionMap.has(rowNumber)) {
+                        let rowSelection = rowIdToSelectionMap.get(rowNumber);
+                        rowSelection.push(range);
+                    } else {
+                        rowIdToSelectionMap.set(rowNumber, [range]);
                     }
+                    rowIdToRowMap.set(rowNumber, row);
                 }
             };
         });
 
+        // get all the rows
         let p = tasks[0]();
         for (let i = 1; i < tasks.length; i++) {
             p = p.then(tasks[i]);
         }
         await p;
+
+        // Go through all rows and get selections for them
+        let allRowIds = rowIdToRowMap.keys();
+        for (let rowId of allRowIds) {
+            let row = rowIdToRowMap.get(rowId);
+            const rowSelections = rowIdToSelectionMap.get(rowId);
+            for (let i = 0; i < rowSelections.length; i++) {
+                let rowSelection = rowSelections[i];
+                for (let j = 0; j < rowSelection.fromCell; j++) {
+                    copyString += ' \t';
+                }
+                let cellObjects = row.slice(rowSelection.fromCell, (rowSelection.toCell + 1));
+                // Remove newlines if requested
+                let cells = this.shouldRemoveNewLines()
+                ? cellObjects.map(x => this.removeNewLines(x.displayValue))
+                : cellObjects.map(x => x.displayValue);
+                copyString += cells.join('\t');
+            }
+            copyString += os.EOL;
+        }
+
+        // Remove any extra lines
+        copyString.trim();
```

Comment:
Again, you can't use trim otherwise you risk removing user content. Just specifically remove the last EOL - we know there'll always be one (because of the header row) and we only care about removing that last one.

Reply:
Good point, didn't consider that.